test/test_connectionpool.py

```diff
@@ -308,7 +311,80 @@ def kaboom(*args, **kwargs):
-
+    
+    def test_preserves_request_header_order(self):
+        """
+        Ensure that headers provided to ConnectionPool.urlopen()
+        are sent to the underlying httplib in the same order.
+        """
+        class AbortRequest(BaseException):
+            pass
+        
+        expected_request_headers = OrderedDict([('X-Header-%d' % i, str(i)) for i in range(16)])
+        
+        c = connection_from_url('http://localhost:80')
+        
+        # Defer: Assert the request sent to httplib had the same header order
+        #        as the original header arguments to ConnectionPool.urlopen().
+        def patched_request(*args, **kwargs):
+            actual_request_headers = kwargs['headers']
+            self.assertEqual(
+                list(actual_request_headers.items()),
+                list(expected_request_headers.items()))
+            raise AbortRequest()
+        
+        # Patch the .request() for new connections in the pool 
+        original_new_conn = c._new_conn
+        def patched_new_conn(*args, **kwargs):
+            conn = original_new_conn(*args, **kwargs)
+            conn.request = patched_request
+            return conn
+        c._new_conn = patched_new_conn
+        
+        try:
+            c.urlopen('GET', '/', headers=expected_request_headers, release_conn=False)
+        except AbortRequest:
+            pass
+    
+    def test_preserves_response_header_order(self):
+        """
+        Ensure that headers returned by ConnectionPool.urlopen()
+        are in the same order as received from the underlying httplib.
+        """
+        
+        expected_response_headers = OrderedDict([('X-Header-%d' % i, str(i)) for i in range(16)])
+        
+        if six.PY3:
+            httplib_response_headers = HTTPMessage()
+        else:
+            httplib_response_headers = HTTPMessage(BytesIO(b''))
+        for (k, v) in expected_response_headers.items():
+            httplib_response_headers.add_header(k, v)
+        
+        class FakeHTTPResponse(object):
+            status = 200
+            reason = 'OK'
+            version = '0.9'
+            length = 0
+            msg = httplib_response_headers
+        httplib_response = FakeHTTPResponse()
+        
+        c = connection_from_url('http://localhost:80')
+        
+        # Patch the .getresponse() for new connections in the pool to return
+        # a simulated httplib.HTTPResponse
+        original_new_conn = c._new_conn
+        def patched_new_conn(*args, **kwargs):
+            conn = original_new_conn(*args, **kwargs)
+            conn.request = lambda *args, **kwargs: None
+            conn.getresponse = lambda *args, **kwargs: httplib_response
+            return conn
+        c._new_conn = patched_new_conn
+        
+        actual_response = c.urlopen('GET', '/', release_conn=False)
+        self.assertEqual(
+            list(actual_response.headers.items()),
+            list(expected_response_headers.items()))
```

Comment:
@davidfstr you're duplicating a lot of code between these tests when you could instead write a different type of test (a socket-level test) with the dummyserver that would require a lot less work in each test, be easier to read, understand, and maintain.